src/hydro/srcterms/tabular_cooling.cpp

```diff
@@ -460,7 +456,11 @@ void TabularCooling::SubcyclingFixedIntSrcTerm(MeshData<Real> *md, const Real dt
           sub_iter++;
         }
 
-        PARTHENON_REQUIRE(internal_e > internal_e_floor, "cooled below floor");
+        // If cooled below floor, reset to floor value.
+        // This could happen if the floor value is larger than the lower end of the
+        // cooling table or if they are close and the last subcycle in the cooling above
+        // the lower end pushed the temperature below the lower end (and the floor).
+        internal_e = (internal_e > internal_e_floor) ? internal_e : internal_e_floor;
```

Comment:
So this would allow cooling below the cooling table to occur if the cooling at the end of the table is fast. Should the temperature instead be fixed to the bottom of the cooling table?

Note that there is already a mechanism above to stop gas that is already below `internal_e_floor` from doing any cooling.

Maybe we should introduce a `cooling_internal_e_floor` and `global_internal_e_floor` below which, respectively, no cooling occurs and below which the internal energy is raised to match the global floor.

Reply:
As I mentioned above, my intention here was that if we end up below the "global" temperature floor we should already apply it here and not first in the ConsToPrim update.
In practice this will likely have no significant impact.

If you agree that this approach is what we want then I should probably also update the Townsend cooling as for that cooling I do not have this check at the end (I only check for the low end of the cooling table).

Regarding a split between `cooling_internal_e_floor` and `global_internal_e_floor`, I'd prefer to leave it at the single, global parameter approach and that the cooling floor is implicitly handled by the lower end of the cooling table.
What do you think?

Reply:
The distinction between the two floors is that the bottom of the cooling table represents something physical and the `Tfloor` is something numerical. As I understand the cooling tables I'm using, cooling should stop at the floor of the cooling table since the cooling time at that point is much longer than the timescale of the simulation. `Tfloor` might be lower than the floor of the cooling table and other processes might lead to lower temperatures, but cooling itself shouldn't drop the temperature below the floor of the cooling table as that would be unphysical.

I think this confusion arose because originally when I wrote my cooling integrator, the only temperature floor I considered was the floor of the cooling table. The temperature floor for the fluid solver incorrectly supplanted that floor within the cooling function. For the Enzo-like runs this hasn't mattered since the floor to the fluid solver was set to the same floor as the cooling table.

Reply:
I don't think either are physical. The cooling tables go down to 10^4 K, right? Ly-a cooling peaks there, so the cooling timescale should be extremely short.